prow/cmd/branchprotector/README.md

```diff
@@ -154,6 +154,13 @@ So in the example above:
     - Enable protection (inherited from branch-protection level)
     - Require the `cla` context to be green to merge (appended by parent)
 
+
+#### Unmanaged branch protection
+
+If it is not desired to have branchprotector manage the branch protection, use 
+`unmanaged: true` in the configuration instead of `protect: false` which 
+disables branch protection. 
```

Comment:
One limitation that `unmanaged: true` has (I did not realize that, just now as I was looking into it) is that it cannot be overriden. Hence it does not  (easily) allow the legit and I guess common use case where you do not want to protect any branches except one (say, the `main` one). You'd expect to be able to set `unamanged: true` on the `repo level and then override it with `unmanaged: false` (?) on branch level, but you cannot do that.

Not sure if this is a blocker or not but I guess we need to consider that case.

Reply:
I saw "Lower levels can not override it." on [our doc](https://docs.ci.openshift.org/docs/architecture/branch-protection/) too.

I guess for the case of "don't touch my branch protection setting except that I want to protect the main branch":
a ugly (we have to explicitly list x ... z) workaround is that 

```yamal
branches:
  main:
    protected: true
  x:
    unmanged: true
  ...
  z:
    unmanged: true
```

Let me ask Alraro why it cannot be overridden.

Reply:
The original motivation for adding `unmanaged` was to avoid making any api requests into orgs or repos that have this set. As soon as we traverse into an org, we [might request all its repos from the github api](https://github.com/kubernetes/test-infra/blob/9703a3111217af8a9a4616883ab316c28f33e3cf/prow/cmd/branchprotector/protect.go#L262) and as soon as we traverse into a repo, we [fetch it from GitHub](https://github.com/kubernetes/test-infra/blob/9703a3111217af8a9a4616883ab316c28f33e3cf/prow/cmd/branchprotector/protect.go#L305). @petr-muller even expressed his liking of that, together with the concern that it might eventually surprise ppl: https://github.com/kubernetes/test-infra/pull/21999#pullrequestreview-647381623

That being said, I image it should be possible to add a check if we have more granular settings and only skip traversing down if that is not the case.

Reply:
> it should be possible to add a check if we have more granular settings and only skip traversing down if that is not the case.

Let me give it a try.

Reply:
Tried thinking ... And my brain got OOMKilled very soon after. ^_^
I cannot get a good understanding of `unmanaged: false` and it is even harder to combine `unmanaged` with `protect`. Both of the has 3 possible values: true, false and nil.

So I took one step back:
Revisit the case: "don't touch my branch protection setting except that I want to protect the main branch".
It is true we cannot do the case easily but it has been like it even without this PR, right?

About the ugly workaround above:
Suppose a repo owner forgot to add `unmanged: true` for a new branch in the config in the above case.
Most likely, `protect` is missing to for the branch.
Then the worst case scenario is that the repo owner might have to set up the bp setting again after adding `unmanged: true`.
I **think** not a very big deal.

Reply:
I can not say how big a deal or not this is, but the required change would be to:
* Add a func that can check if any branch on a repo has a non-nil `true` setting of `protect`
* Extend the [check here](https://github.com/kubernetes/test-infra/blob/9703a3111217af8a9a4616883ab316c28f33e3cf/prow/cmd/branchprotector/protect.go#L301) to call this func, so it will look something like `if repo.Policy.Unmanaged != nil && *repo.Policy.Unmanaged && !repo.HasManagedBranches()`

And something similiar for the org setting.

If you do this, it is likely better to do in a separate change though.

Reply:
Thanks for the guidance. Let me give another try.

> If you do this, it is likely better to do in a separate change though.

Understood.

Reply:
Resolved by https://github.com/kubernetes/test-infra/pull/25456